debezium/dbz#1930 Add Debezium Source Connector Development Guide (#7495)
debezium/debezium: documentation/modules/ROOT/pages/development/source-connector.adoc

[cols="2,4,5",options="header"]
|===
| Aspect | Plain Kafka Connect source connector | Source connector built on the {prodname} framework

| Base API
| `SourceConnector` / `SourceTask`
| `BaseSourceConnector` / `BaseSourceTask`, which extend the Kafka Connect classes

| Lifecycle
| You implement `start()`, `poll()`, and `stop()` from scratch
| You still implement `start(Configuration)`, `doPoll()`, and `doStop()`, but `BaseSourceTask` provides the surrounding template (validation, restart backoff, offset commits) and `ChangeEventSourceCoordinator` manages the snapshot-to-streaming handover

| Offset management
| Hand-built maps attached to each `SourceRecord`
| Structured through `Partition`, `OffsetContext`, and `OffsetContext.Loader`

| Schema and records
| Hand-built `Schema` and `Struct` objects
| `DatabaseSchema` caches per-table schemas; the `Envelope` factory builds change event values with `source` metadata

| Snapshot and streaming
| Your own ad-hoc logic
| Explicit `SnapshotChangeEventSource` and `StreamingChangeEventSource` contracts, coordinated by the framework

| Envelope format
| No envelope required

> | Standard {prodname} envelope (`op`, `before`, `after`, `source`, `ts_ms`)

Reviewer: I wonder if it would be good to mention here that different message types have different envelope structures. For example, a heartbeat or schema change event looks very different from a change event, but each has its own envelope structure.

| Error handling
| Entirely manual
| Still implemented by the connector, with `ErrorHandler` as a helper that classifies retriable versus fatal failures and drives task restarts
|===

The runtime is not a differentiator.

> Kafka Connect runs any source connector, and xref:{link-debezium-server}[{prodname} Server] and the xref:{link-engine}[{prodname} Engine] embed the Kafka Connect API, so they can also run source connectors.

Reviewer: ```suggestion
A {prodname} source connector can run across various runtimes, such as Kafka Connect, xref:{link-debezium-server}[{prodname} Server], or xref:{link-engine}[{prodname} Engine].
```
I'm rephrasing this because I want to avoid the impression that Debezium Server or Engine embeds the Kafka Connect API, etc. We're actively trying to move away from that where possible.

The practical benefit of implementing the {prodname} contracts is that your events carry the standard envelope, which makes them work with {prodname}'s converters, single message transformations, and sink ecosystem.

== Architecture overview

> === Component overview

Reviewer: I realize this is likely generic, but nowhere do we raise JdbcConnection or have the `MyConnection` subclass. This should at least be mentioned in the context of relational connectors, since that is generally where most "query" or "database interfacing" should be routed.

    │   └── MyStreamingChangeEventSource   (StreamingChangeEventSource)
    ├── MyChangeRecordEmitter              (ChangeRecordEmitter)
    ├── MyErrorHandler                     (ErrorHandler)
    └── Framework services instantiated in the task:
        ChangeEventQueue, EventDispatcher, ChangeEventSourceCoordinator,
        SignalProcessor, NotificationService, SnapshotterService
....

The classes on the left are yours.
The services at the bottom come from `debezium-connector-common` and are only instantiated and wired together in your task class.

=== Runtime flow

At runtime, every {prodname} connector follows the same five-step flow:

. Kafka Connect (or {prodname} Server) instantiates the connector class, validates the configuration, and creates the task configurations.
. The task starts, loads previously persisted offsets through the `OffsetContext.Loader`, and builds the pipeline: schema, queue, dispatcher, and coordinator.
. The `ChangeEventSourceCoordinator` asks the `Snapshotter` whether a snapshot is needed and, if so, runs the `SnapshotChangeEventSource`, which emits `READ` (`r`) events for existing data.

> . When the snapshot completes, the coordinator starts the `StreamingChangeEventSource` at the offset where the snapshot finished, and streaming emits `CREATE` (`c`), `UPDATE` (`u`), and `DELETE` (`d`) events from that point on.

Reviewer: ```suggestion
. When the snapshot completes, the coordinator starts the `StreamingChangeEventSource` at the database position where the snapshot was taken, and streaming emits `CREATE` (`c`), `UPDATE` (`u`), and `DELETE` (`d`) events from that point on.
```

> . Emitted events flow through the `ChangeEventQueue` into the Kafka Connect polling loop, and Kafka Connect periodically persists the offsets that your `OffsetContext.getOffset()` returns.

Reviewer: ```suggestion
. Emitted events flow through the `ChangeEventQueue` into the runtime's polling loop. Periodically, the runtime persists the offsets returned by your `OffsetContext.getOffset()`.
```

=== The connector configuration

Your configuration class defines the connector's `Field` constants and gives the framework typed access to common settings.
Which base class you extend is the first consequence of the design decision from the architecture overview:

* `CommonConnectorConfig` — sources that are not relational (MongoDB).
* `RelationalDatabaseConnectorConfig` — relational sources without schema history (PostgreSQL).
* `HistorizedRelationalDatabaseConnectorConfig` — relational sources with schema history (SQL Server, MySQL, Oracle).

[source,java]
----
public class MyConnectorConfig extends RelationalDatabaseConnectorConfig {

    public static final Field MY_PROPERTY = Field.create("my.property")
            .withType(ConfigDef.Type.STRING)
            .withImportance(ConfigDef.Importance.HIGH)
            .withDescription("Description shown in the Connect UI and generated docs.")
            .required();

    @Override
    protected SourceInfoStructMaker<?> getSourceInfoStructMaker(Version version) {
        return new MySourceInfoStructMaker();
    }
}
----

> `CommonConnectorConfig` already defines the connector-agnostic options that every {prodname} connector inherits: `topic.prefix`, `max.queue.size`, `max.batch.size`, `poll.interval.ms`, `tombstones.on.delete`, signal and notification channels, and more.

Reviewer 2: I would also add something on the `ConfigDefinition` where all config are grouped.

=== Supporting services

[cols="2,4,4",options="header"]
|===
| Service | Purpose | Examples in the main repository

| `ErrorHandler`
| Classifies errors as retriable or fatal when the streaming loop fails.
Override `communicationExceptions()` (or `isRetriable(...)`) to declare which exceptions warrant a task restart.
| `PostgresErrorHandler` adds `SQLException` to the default `IOException`; `SqlServerErrorHandler` likewise adds `SQLException`

| `EventMetadataProvider`
| Extracts event timestamp, position, and transaction id from emitted events, feeding metrics, heartbeats, and the transaction monitor.
| `SqlServerEventMetadataProvider`

| `ChangeEventSourceMetricsFactory`
| Creates the snapshot and streaming JMX metrics.
Use `DefaultChangeEventSourceMetricsFactory` unless you need custom metrics.
| `SqlServerMetricsFactory`

| Value converters
| Map source data types to Kafka Connect schema types, including the `decimal.handling.mode` and `time.precision.mode` semantics.
Relational connectors extend `JdbcValueConverters`.

> | `SqlServerValueConverters`, `PostgresValueConverter`

Reviewer: I think that in the section where we talk about the files you should create when building your own connector, we should include an example of `MyValueConverters` that, for relational connectors, extends `JdbcValueConverters`. 

We also did not touch on `DefaultValueConverter`, which is the sibling to the ValueConverters but specifically for converting column default value metadata.

NOTE: In {prodname} versions before the module rename, the framework artifact is `io.debezium:debezium-core`.
Depending on `debezium-core` still works, because the artifact relocates to `debezium-connector-common`, but new connectors should depend on `debezium-connector-common` directly.

Package the connector with the Maven Shade or Assembly plugin so that connector-specific dependencies are included in the plugin directory, while provided runtime dependencies are not.

=== Package structure and conventions

Follow the `io.debezium.connector.<name>` package convention used by all {prodname} connectors, with one class per framework contract as shown in the component overview.

Two resource conventions matter:

* `META-INF/services/org.apache.kafka.connect.source.SourceConnector` must list your connector class so that the Kafka Connect plugin scanner discovers it.
* Add a `Module` class and a Maven-filtered `build.version` properties file, so that `version()` reports the project version; see `io.debezium.connector.sqlserver.Module` for the pattern.

== Implementation walkthrough

This section builds each contract in dependency order.

> Every subsection states the contract, shows a short generic snippet, and points at the SQL Server and PostgreSQL implementations for the production-grade details.

Reviewer 3: Add a note explaining that connector-related classes should be prefixed with the source name, following the established naming convention. Use an example source name such as `AcmeDB` and consistently reference it throughout all code snippets and examples instead of using `my`.

=== SourceInfo and SourceInfoStructMaker

Every change event carries a `source` struct that identifies where the event came from.
You model the mutable state in a `SourceInfo` class extending `io.debezium.connector.common.BaseSourceInfo`, and you describe its schema in a `SourceInfoStructMaker`, usually by extending `AbstractSourceInfoStructMaker`, which contributes the common fields (`version`, `connector`, `name`, `ts_ms`, `snapshot`, `db`, and so on):

[source,java]
----
public class MySourceInfoStructMaker extends AbstractSourceInfoStructMaker<MySourceInfo> {

    private Schema schema;

    @Override
    public void init(String connector, String version, CommonConnectorConfig config) {
        super.init(connector, version, config);
        schema = commonSchemaBuilder()
                .name("io.debezium.connector.my.Source")
                .field("position", Schema.INT64_SCHEMA)   // your position fields
                .build();
    }

    @Override
    public Schema schema() {
        return schema;
    }

    @Override
    public Struct struct(MySourceInfo sourceInfo) {
        return commonStruct(sourceInfo).put("position", sourceInfo.getPosition());
    }
}
----

> For comparison, SQL Server's `SourceInfo` carries `change_lsn`, `commit_lsn`, and `event_serial_no`, and PostgreSQL's carries `lsn`, `txId`, and `xmin`.

Reviewer 3: add this as a note

=== Partition and Partition.Provider

A `Partition` identifies one independently-captured slice of the source and provides the map that Kafka Connect uses as the source partition key for offset storage.
Most connectors have exactly one partition per task, keyed by the logical server name:

[source,java]
----
public class MyPartition implements Partition {

    @Override
    public Map<String, String> getSourcePartition() {
        return Map.of("server", serverName);
    }

    // Implement equals() and hashCode(): partitions are used as
    // map keys by the coordinator and the metrics.
}
----

The nested `Partition.Provider` returns the set of partitions the task must process, typically a singleton set.

> SQL Server is the instructive production example: its `SqlServerPartition.Provider` returns one partition per configured database (`{"server": ..., "database": ...}`), which is what enables its multi-database mode.

Reviewer 3: again as a note

`RelationalSnapshotChangeEventSource` already implements the entire snapshot algorithm (determining captured tables from the include/exclude filters, locking, reading table structure, persisting schema history, and reading data in chunks) and leaves you a handful of template methods:

[source,java]
----
protected abstract Set<TableId> getAllTableIds(RelationalSnapshotContext<P, O> ctx);
protected abstract void lockTablesForSchemaSnapshot(ChangeEventSourceContext ctx, RelationalSnapshotContext<P, O> snapshotCtx);
protected abstract void determineSnapshotOffset(RelationalSnapshotContext<P, O> ctx, O previousOffset);
protected abstract void readTableStructure(ChangeEventSourceContext ctx, RelationalSnapshotContext<P, O> snapshotCtx, O offset, SnapshottingTask task);
protected abstract void releaseSchemaSnapshotLocks(RelationalSnapshotContext<P, O> ctx);
protected abstract SchemaChangeEvent getCreateTableEvent(RelationalSnapshotContext<P, O> ctx, Table table);
protected abstract Optional<String> getSnapshotSelect(RelationalSnapshotContext<P, O> ctx, TableId tableId, List<String> columns);
protected abstract O copyOffset(RelationalSnapshotContext<P, O> snapshotContext);
----

> See `SqlServerSnapshotChangeEventSource` and `PostgresSnapshotChangeEventSource` for complete implementations of these methods.

Reviewer 3: again note

Setting `snapshot.mode` to `custom` selects a user-provided `Snapshotter`, registered the same way and named by `snapshot.mode.custom.name`.

=== The streaming change event source

The `StreamingChangeEventSource` carries most of the connector-specific logic: it obtains changes from the source, updates the offset context, and dispatches one emitter per change.
Its `execute(...)` method must loop until `context.isRunning()` turns false:

[source,java]
----
@Override
public void execute(ChangeEventSourceContext context, MyPartition partition, MyOffsetContext offsetContext)
        throws InterruptedException {
    while (context.isRunning()) {
        for (SourceChange change : pollChanges(offsetContext.position())) {   // your source access
            offsetContext.advanceTo(change.position());
            dispatcher.dispatchDataChangeEvent(partition, change.collectionId(),
                    new MyChangeRecordEmitter(partition, offsetContext, change));
        }
        pauseUntilNewChanges();   // poll interval, watch API, blocking read, ...
    }
}
----

> Dispatching goes through the framework's `EventDispatcher`, never directly to the queue.

Reviewer 3: I was thinking that this section could also be illustrated with a diagram. Do we have a recommended tool for creating diagrams in the documentation, such as [Mermaid](https://mermaid.js.org/)? @Naros

Author: Added a SVG flow diagram similar to other docs instead of mermaid.

The dispatcher applies the collection filter, looks up the schema, runs your emitter, and enqueues the resulting `SourceRecord` into the `ChangeEventQueue`, which applies backpressure when Kafka or the sink is slow.

> For production examples of the two streaming styles, read these two classes side by side:

Reviewer 3: from 562 to 567 can be a note

. Build the connector archive with `mvn clean package`.

. Place the connector JAR (with its bundled dependencies) in a directory on the Kafka Connect `plugin.path`.

. Start Kafka and Kafka Connect, for example with the {prodname} container image:
+
[source,yaml,subs="attributes+"]
----
services:
  kafka:
    image: apache/kafka:latest
  connect:
    image: quay.io/debezium/connect:{debezium-docker-label}
    ports:
      - "8083:8083"
    environment:
      - BOOTSTRAP_SERVERS=kafka:9092
      - GROUP_ID=1
      - CONFIG_STORAGE_TOPIC=my_connect_configs
      - OFFSET_STORAGE_TOPIC=my_connect_offsets
      - STATUS_STORAGE_TOPIC=my_connect_status
    volumes:
      - ./target/my-connector:/kafka/connect/my-connector
----

. Register the connector through the REST API:
+
[source,json]
----
{
    "name": "my-connector",
    "config": {
        "connector.class": "io.debezium.connector.my.MyConnector",
        "tasks.max": "1",
        "topic.prefix": "myserver",
        "my.property": "value"
    }
}
----
+
[source,bash]
----
curl -i -X POST -H "Accept:application/json" -H "Content-Type:application/json" \
  http://localhost:8083/connectors/ -d @register-my-connector.json
----

> === Deploying to {prodname} Server

Reviewer 3: I was thinking that the verification can be done with the Debezium CLI with a new command `run` which can use the `debezium-quarkus-engine`. WDYT @Naros ? 

cc: @Binayak490-cyber

| `connectorClass()`, `config()`, `connector()`, `server()`, expected counts
|===

This is exactly how the production connectors are tested.
SQL Server's `IncrementalSnapshotIT` is a good template; it inherits every scenario and supplies only configuration and dialect:

[source,java]
----
public class IncrementalSnapshotIT extends AbstractIncrementalSnapshotWithSchemaChangesSupportTest<SqlServerConnector> {

    @Override
    protected Class<SqlServerConnector> connectorClass() {
        return SqlServerConnector.class;
    }

    @Override
    protected String topicName() {
        return "server1.testDB1.dbo.a";
    }

    @Override
    protected Builder config() {
        return TestHelper.defaultConfig()
                .with(SqlServerConnectorConfig.SNAPSHOT_MODE, SnapshotMode.NO_DATA)
                .with(SqlServerConnectorConfig.SIGNAL_DATA_COLLECTION, "testDB1.dbo.debezium_signal")
                .with(SqlServerConnectorConfig.INCREMENTAL_SNAPSHOT_CHUNK_SIZE, 250)
                .with(SqlServerConnectorConfig.INCREMENTAL_SNAPSHOT_ALLOW_SCHEMA_CHANGES, true);
    }
    // + signalTableName(), connector(), server(), DDL dialect adapters
}
----

> Postgres's `BlockingSnapshotIT` and `PostgresChunkedSnapshotIT` show the same pattern for the other bases.

Reviewer 3: note
